### PR TITLE
SI-8507 Avoid lazy val in StringContext

### PR DESCRIPTION
**Pull request by som-snytt**
This mild refactor eliminates the lazy val StringBuilder
and arguably makes the code easier to read.

There is a small unit test but no benchmark to prove
anything useful about performance.

### CODE REVIEW

File: src/library/scala/StringContext.scala [hunk 1]
```diff
@@ -167,6 +167,10 @@ case class StringContext(parts: String*) {
 
 object StringContext {
 
+  import annotation.tailrec
+
+  private type StringBuilder = java.lang.StringBuilder
```
Comment (retronym) [inline src/library/scala/StringContext.scala]: Why not an import?

Comment (som-snytt) [inline src/library/scala/StringContext.scala]: Sometimes you just want to drink something domestic.

There's also a stray require.  The stringbuilder param needn't be in the loop. I'll give it another look after tests pass.

**Comment by som-snytt** [comment]
Expanded the exception message:

```
scala> s"\d".r
scala.StringContext$InvalidEscapeException: invalid escape '\d' not one of [\b, \t, \n, \f, \r, \\, \", \'] at index 0 in "\d". Use \\ for literal \.

scala> s"a\"
scala.StringContext$InvalidEscapeException: invalid escape at terminal index 1 in "a\". Use \\ for literal \.
```

FTR, the first time I wrote the string, I didn't use enough \  in "Use \" and it recursed and I straight cursed.

**Comment by som-snytt** [comment]
Maybe @paulp would care to review the early warning system for the escaped quote pitfall in 9d5ee50.  It tries to warn in parser, before you're actually in the pit.

Is that safe because you can never hijack the `s""` and `f""` syntax?

http://osrc.dfm.io/som-snytt

> There is also an uncanny similarity between som-snytt's activity stream and those of Paul Phillips, ...
> 
> It seems like som-snytt is—or should be—friends with Jason Zaugg. With this in mind, it's worth noting that Jason spends less of their time commenting on issues. 

Yeah, I noticed that, too.  What's up with that?  It's like you don't have time for your friends anymore.  Maybe @retronym could take a second look. I addressed the previous comment, and threw in $" under -Xfuture, which seems safe.

**Comment by paulp** [comment]
Don't you know that only absolute generality is acceptable. If the parser knows about the s and f interpolators then the terrorists have won.

To the matter of uncanny similarity, has anyone ever seen som-snytt and paulp at the same time?

**Comment by som-snytt** [comment]
"There may be an infinity of interpolators possible, but most people will only ever use one."

https://groups.google.com/forum/#!msg/scala-internals/_DnZGO-4NXc/xNoiOUoCf58J

Where the one is, of course, the f macro.

`f` is for fastTrack (tm).

The too-bad is that the macro doesn't get a chance to error-out on `f"abc\"def"`.  For arbitrary macro, you could imagine an attempt to recover from the parse error and use partial info to locate the macro and do some validation. 

Practically, though: the band-aid or nicotine patch relieves much suffering.

My first shot was to try to detect the bad escape in Scanner when erroring on the missing quote (in two places). That was more general, more heuristic, probably flakier.  I didn't like that I couldn't just consult the token stack anymore; but maybe I should have just dipped into the source line. That might have sufficed in an 80/20 way.

**Comment by som-snytt** [comment]
Forcibly pushed with test updates. SI-6631 tests moved to junit.  Wow, tests!

**Comment by retronym** [comment]
You can actually hijack `s` and `f`, slightly to my chagrin, by:

``` scala
scala> object AllYourStringsAreBelongToMe { case class StringContext(args: Any*) { def s(args: Any) = "!!!!" } }
defined module AllYourStringsAreBelongToMe

scala> { import AllYourStringsAreBelongToMe._; println(s"Interpolate $this!") }
!!!!
```

**Comment by som-snytt** [comment]
Agh.

I would have said, "Agh!!!!", but for the physical shock.  Is chagrin French for, That makes me ache.

On the upside, specs2 doesn't have to call its macro `s2""` anymore, so now specs2 can also call itself plain specs again.

**Comment by som-snytt** [comment]
To plug the exploit, may I just push:

```
scala> q"""s"hi""""
res1: reflect.runtime.universe.Tree = _root_.scala.StringContext("hi").s()
```

**Comment by retronym** [comment]
I was told that it was a conscious choice to leave this open for language virtualization needs.

**Comment by som-snytt** [comment]
So basically, after all this time, I still can't tell when paulp is being sarcastic. I mean, when he's _not_.

> only absolute generality is acceptable

**Comment by som-snytt** [comment]
@retronym I'll have it emit the helpful message as a supplement to the "unclosed string literal" error.  It never hurts to add info to an error (modulo reams of stack trace).

That requires a tiny bit of coordination between Parser (which knows the interpolation) and Scanner (which sees the error).  Hopefully less clunky than `unit.hasXml` and `unit.firstXmlPos`.

**Comment by som-snytt** [comment]
The improvement is modelled as a bit of parenthetical advice which is appended to a subsequent syntax error that occurs within a line.  The advice could be misguided only if the standard interpolators are hijacked (as described).  A goal of the change is not to impact parser performance by undue cleverness.

**Comment by retronym** [comment]
I'm retiring master in favour of our new `2.{11,12}.x` regime. This PR has sort of grown in scope beyond what I've had time to properly review. Can you please resubmit against the 2.11.x branch?

**Pull request closed without merging**